sourced/ml/mining/cmd/clickhouse2deps.py

```diff
@@ -0,0 +1,110 @@
+import logging
+import os
+from pathlib import Path
+
+from clickhouse_driver import Client
+import jinja2
+from scipy.sparse import coo_matrix
+import yaml
+
+from sourced.ml.mining.models import Dependencies
+
+
+QUERY_TEMPLATE = "clickhouse2deps.sql.jinja2"
+QUERY_ARGS = "clickhouse2deps.yaml"
+CLICKHOUSE_LANGS = [
+    "cpp",
+    "csharp",
+    "go",
+    "java",
+    "javascript",
+    "php",
+    "python",
+]  # TODO(r0mainK): add ruby
+MAX_BLOCK_SIZE = 1000000
+
+
+def clickhouse2deps(args):
+    """
+    Extract dependencies from UASTs in a Clickhouse DB.
+    """
+    log = logging.getLogger("clickhouse2deps")
+    if not args.output_path.suffix == ".asdf":
+        args.output_path = args.output_path.with_suffix(".asdf")
+    if args.output_path.exists():
+        if not args.force:
+            log.error(
+                "%s already exists, aborting (use -f/--force to overwrite)",
+                args.output_path,
+            )
+            raise FileExistsError
+        log.warn("%s already exists, overwritten", args.output_path)
+        os.remove(args.output_path)
+    log.info("Loading the query template ...")
+    root = Path(__file__).parent
+    template_loader = jinja2.FileSystemLoader(str(root))
+    env = jinja2.Environment(keep_trailing_newline=False)
+    template = template_loader.load(env, name=QUERY_TEMPLATE)
+    log.info("Loading the query args ...")
+    with open(root / QUERY_ARGS, "r") as fin:
+        query_args = yaml.load(fin)
+    client = Client(
+        user=args.user,
+        password=args.password,
+        host=args.host,
+        port=args.port,
+        database=args.database,
+    )
+    files, deps, ind_files, ind_deps = [], [], [], []
+    ind_to_langs, ind_to_repos, file_to_inds, dep_to_inds = {}, {}, {}, {}
+    for lang in args.langs:
+        log.info("Extracting %s dependencies...", lang)
+        rows = client.execute_iter(
+            template.render(lang=lang, table=args.table, query_args=query_args[lang]),
+            settings={"max_block_size": MAX_BLOCK_SIZE},
+        )
+        num_rows, num_files, num_deps = 0, 0, 0
+        for row in rows:
+            row = [
+                e.decode("utf-8", errors="ignore") if not isinstance(e, str) else e
+                for e in row
```

Comment:
I wonder: which are the cases when it is not a string?

Reply:
it happens 4 times in the present DB, all for C++, I can track them down tonight